File: client/src/main/java/au/com/codeka/warworlds/client/ctrl/ExpandableStarListAdapter.java
package au.com.codeka.warworlds.client.ctrl;


import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;

import javax.annotation.Nonnull;

import au.com.codeka.warworlds.client.App;
import au.com.codeka.warworlds.client.concurrency.Threads;
import au.com.codeka.warworlds.client.game.world.StarCollection;
import au.com.codeka.warworlds.client.util.eventbus.EventHandler;
import au.com.codeka.warworlds.common.proto.Star;

/**
 * Base adapter class for adapter which show an expandable list of stars which you can expand and
 * collapse and show other things inside of.
 *
 * We assume the children we care about are all of type {@code T}.
 */
public abstract class ExpandableStarListAdapter<T> extends BaseExpandableListAdapter {
  private final StarCollection stars;

  public ExpandableStarListAdapter(@Nonnull StarCollection stars) {
    this.stars = stars;
    App.i.getEventBus().register(eventListener);
  }

  /**
   * Destroy this {@link ExpandableStarListAdapter}. Must be called to unregister us from the event
   * bus when you're finished.
   */
  public void destroy() {
    App.i.getEventBus().unregister(eventListener);
  }

  @Override
  public boolean hasStableIds() {
    return true;
  }

  @Override
  public boolean isChildSelectable(int groupPosition, int childPosition) {
    return true;
  }

  @Override
  public int getGroupCount() {
    return stars.size();
  }

  @Override
  public Star getGroup(int groupPosition) {
    return getStar(groupPosition);
  }

  protected abstract int getNumChildren(Star star);
  protected abstract T getChild(Star star, int index);
  protected abstract View getStarView(Star star, View convertView, ViewGroup parent);
  protected abstract View getChildView(
      Star star, int index, View convertView, ViewGroup parent);
  protected abstract long getChildId(Star star, int childPosition);

  @Override
  public int getChildrenCount(int groupPosition) {
    Star star = getStar(groupPosition);
    if (star == null) {
      return 0;
    }

    return getNumChildren(star);
  }

  @Override
  public T getChild(int groupPosition, int childPosition) {
    Star star = getStar(groupPosition);
    if (star == null) {
      return null;
    }

    return getChild(star, childPosition);
  }

  @Override
  public long getGroupId(int groupPosition) {
    return stars.get(groupPosition).id;
  }

  @Override
  public long getChildId(int groupPosition, int childPosition) {
    return getChildId(stars.get(groupPosition), childPosition);
  }

  @Override
  public View getGroupView(int groupPosition, boolean isExpanded, View convertView,
      ViewGroup parent) {
    Star star = getStar(groupPosition);
    return getStarView(star, convertView, parent);
  }

  @Override
  public View getChildView(int groupPosition, final int childPosition, boolean isLastChild,
      View convertView, ViewGroup parent) {
    Star star = getStar(groupPosition);
    return getChildView(star, childPosition, convertView, parent);
  }

  protected Star getStar(int index) {
    return stars.get(index);
  }

  private final Object eventListener = new Object() {
    @EventHandler(thread= Threads.UI)
    public void onStarUpdated(Star star) {
      if (stars.notifyStarModified(star)) {
        notifyDataSetChanged();
      }
    }
  };
}

File: client/src/main/java/au/com/codeka/warworlds/client/ctrl/ExpandableStarListAdapter.kt
package au.com.codeka.warworlds.client.ctrl

import android.view.View
import android.view.ViewGroup
import android.widget.BaseExpandableListAdapter
import au.com.codeka.warworlds.client.App
import au.com.codeka.warworlds.client.concurrency.Threads
import au.com.codeka.warworlds.client.game.world.StarCollection
import au.com.codeka.warworlds.client.util.eventbus.EventHandler
import au.com.codeka.warworlds.common.proto.Star

/**
 * Base adapter class for adapter which show an expandable list of stars which you can expand and
 * collapse and show other things inside of.
 *
 * We assume the children we care about are all of type `T`.
 */
abstract class ExpandableStarListAdapter<T>(private val stars: StarCollection):
    BaseExpandableListAdapter() {

  /**
   * Destroy this [ExpandableStarListAdapter]. Must be called to unregister us from the event
   * bus when you're finished.
   */
  fun destroy() {
    App.i.eventBus.unregister(eventListener)
  }

  override fun hasStableIds(): Boolean {
    return true
  }

  override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean {
    return true
  }

  override fun getGroupCount(): Int {
    return stars.size()
  }

  override fun getGroup(groupPosition: Int): Star {
    return getStar(groupPosition)
  }

  protected abstract fun getNumChildren(star: Star?): Int
  protected abstract fun getChild(star: Star?, index: Int): T
  protected abstract fun getStarView(star: Star?, convertView: View?, parent: ViewGroup?): View
  protected abstract fun getChildView(
      star: Star?, index: Int, convertView: View?, parent: ViewGroup?): View

  protected abstract fun getChildId(star: Star?, childPosition: Int): Long
  override fun getChildrenCount(groupPosition: Int): Int {
    val star = getStar(groupPosition)
    return getNumChildren(star)
  }

  override fun getChild(groupPosition: Int, childPosition: Int): T {
    val star = getStar(groupPosition)
    return getChild(star, childPosition)
  }

  override fun getGroupId(groupPosition: Int): Long {
    return stars[groupPosition].id
  }

  override fun getChildId(groupPosition: Int, childPosition: Int): Long {
    return getChildId(stars[groupPosition], childPosition)
  }

  override fun getGroupView(groupPosition: Int, isExpanded: Boolean, convertView: View,
                            parent: ViewGroup): View {
    val star = getStar(groupPosition)
    return getStarView(star, convertView, parent)
  }

  override fun getChildView(groupPosition: Int, childPosition: Int, isLastChild: Boolean,
                            convertView: View, parent: ViewGroup): View {
    val star = getStar(groupPosition)
    return getChildView(star, childPosition, convertView, parent)
  }

  protected fun getStar(index: Int): Star {
    return stars[index]
  }

  private val eventListener: Any = object : Any() {
    @EventHandler(thread = Threads.UI)
    fun onStarUpdated(star: Star?) {
      if (stars.notifyStarModified(star)) {
        notifyDataSetChanged()
      }
    }
  }

  init {
    App.i.eventBus.register(eventListener)
  }
}